test/e2e/scalenodes.go

```diff
@@ -75,11 +80,59 @@ var _ = Describe("Scale nodes", func() {
 		err = waitForScale(mss.Items[0].Name)
 		Expect(err).NotTo(HaveOccurred())
 	})
+
+	Specify("operator should maintain at least two worker replicas", func() {
+		mss, err := clients.MachineAPI.MachineV1beta1().MachineSets(machineSetsNamespace).List(context.Background(), metav1.ListOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(mss.Items).NotTo(BeEmpty())
+
+		switch {
+		case len(mss.Items) == 3:
+			// E2E cluster has AZs, remove one replica from 2 machinesets
+			err = scale(mss.Items[0].Name, -1)
+			Expect(err).NotTo(HaveOccurred())
+
+			err = waitForScale(mss.Items[0].Name)
+			Expect(err).NotTo(HaveOccurred())
+
+			err = scale(mss.Items[1].Name, -1)
+			Expect(err).NotTo(HaveOccurred())
+
+			err = waitForScale(mss.Items[1].Name)
+			Expect(err).NotTo(HaveOccurred())
+
+		case len(mss.Items) < 3:
+			// E2E cluster has no AZs, remove two replicas from 1 machineset
+			err = scale(mss.Items[0].Name, -2)
+			Expect(err).NotTo(HaveOccurred())
+
+			err = waitForScale(mss.Items[0].Name)
+			Expect(err).NotTo(HaveOccurred())
+		}
+
+		ms, err := clients.MachineAPI.MachineV1beta1().MachineSets(machineSetsNamespace).List(context.Background(), metav1.ListOptions{})
+		Expect(err).NotTo(HaveOccurred())
+
+		replicaCount := 0
+		for _, machineset := range ms.Items {
+			if machineset.Spec.Replicas != nil {
+				replicaCount += int(*machineset.Spec.Replicas)
+			}
+		}
+		Expect(replicaCount).To(BeEquivalentTo(minSupportedReplicas))
+
+		// revert the cluster to previous state
+		err = scale(mss.Items[0].Name, 1)
+		Expect(err).NotTo(HaveOccurred())
+
+		err = waitForScale(mss.Items[0].Name)
+		Expect(err).NotTo(HaveOccurred())
```

Comment:
In either case, it's `mss.Items[0].Name` that ends up one replica short, so it's the same operation for both.

I haven't tested e2e without that there, just thought it proper to fix the cluster afterwards.

Reply:
In the multi AZ setup, the operator allows e2e to scale down `mss.Items[0].Name` to 0 because a minimum of two total replicas are still met by the other two machinesets. The scale-down of `mss.Items[1].Name` is then reverted by the operator, because total replicas have dipped below 2. So at the end of the test, `mss.Items[1].Name` has 1 replica but `mss.Items[0].Name` is left with 0, so we add the replica back.

In the no-AZ setup, `mss.Items[0].Name` starts with three replicas, and the test removes two, but the operator only adds one back. So in this case, we do the same thing and just add a single replica back to `mss.Items[0].Name`.

If it's not clear from the code, we can probably just leave out lines 124-129.

Reply:
@m1kola I took the easy route for now, and E2E passes. Is the concern that we'd be introducing a flake by allowing E2E to continue with only 2 workers?

Reply:
@cadenmarchese my concern is that

1. People writing e2e might expect 3 worker nodes at the start and might be not aware of this test. So they create PR with a new e2e test and everything is green because their new test gets executed before this one which modifies the nubmer of replicas. But after the merge execution order of e2e tests gets changes for some reason and e2e start failing sometimes.
1. Existing e2es might rely on the fact that there are 3 worker nodes from the beginning. Similar to the previous point - everything might be green now, but what if test execution order changes?

Hopefully it will be not too hard to debug if this actually happens: if someone is relying on a having 3 nodes - then the failure will likely indicate that.

So let's merge and see how it goes.

Reply:
@m1kola sounds good, I will keep an eye on failed e2e runs. Thanks for all the help here, I really appreciate it.